zerver/views/user_settings.py

```diff
@@ -247,13 +247,14 @@ def json_change_settings(
             ):
                 raise JsonableError(_("Wrong password!"))
         except RateLimitedError as e:
```

Comment:
Hmm, I would have thought that this exception would only be possible if `should_rate_limit(request)` -- can you explain under what situation that can happen? 

I would have thought that all the `should_rate_limit` calls would be in code paths before raising a `RateLimitedError`.

Reply:
Our backend for `authenticate` doesn't call `should_rate_limit` at all: modifying it to do so is a next step (and alternative to the commit that added "just swallow the RateLimitedError" behaviors) I mentioned in the PR description: https://github.com/zulip/zulip/blob/92f085c571d5cb284539e8fe1c6b7e5af9f4c676/zproject/backends.py#L294-L304

There's a chance, albeit slim, that I've forgotten something from my hours in `ipdb` and there actually is a call to `should_rate_limit` somewhere in that call graph, but to my recollection, there's not: much as one of these commits removed a few mostly-unconditional rate limit calls, our auth paths seem to have a similar bug.

Reply:
Ah, yeah: eventually that path calls into this block, which unconditionally applies rate limiting (I'm fairly sure - I haven't traced every single implementer of this ABC): https://github.com/zulip/zulip/blob/22712f634d85102613ae1b0adaa678898d65abc8/zerver/lib/rate_limiter.py#L47-L69

Reply:
I'm not entirely sure why or how 6af799e14a74043ac0bff0098166199e627a986f worked at this point honestly: I've found the actual source of the `RateLimitedErrors`, and it was further along in this method than the call to `authenticate` (I'd alluded to a suspicion I was missing something, in DMs earlier: well, this was it). I'm thinking I may have had false-positive successful runs as a blip in testing all this, because indeed: the auth flows _do_ correctly check `should_rate_limit(request)`, a detail I'd forgotten about in the mess of debugging.